File: src/phenotype.erl
%%%-------------------------------------------------------------------
%%% @author borja
%%% @doc
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(phenotype).
-compile([export_all, nowarn_export_all]). %% TODO: To delete after build

%% API
-export([]).
-export_type([id/0, phenotype/0]).

-type id() :: {Reference :: reference(), phenotype}.
-define(PHENOTYPE_ID, {make_ref(), phenotype}).

-record(phenotype, {
    id           :: id(), 
    network      :: enn:id(),
    actuators    :: [actuator:id()],
    sensors      :: [  sensor:id()]
}).
-type phenotype() :: #phenotype{}.


-define(   CORTEX(State), maps:get(   cortex, State)).
-define(ACTUATORS(State), maps:get(actuators, State)).
-define(  SENSORS(State), maps:get(  sensors, State)).


%%%===================================================================
%%% API 
%%%===================================================================

%%--------------------------------------------------------------------
%% @doc Returns the record fields of an element.
%% @end
%%--------------------------------------------------------------------
-spec fields(Atom :: phenotype) -> ListOfFields :: [atom()].
fields(phenotype) -> record_info(fields, phenotype).

%%--------------------------------------------------------------------
%% @doc Returns the id of a new phenotype.
%% @end
%%--------------------------------------------------------------------
-spec new_from(Genotype :: genotype:id()) -> id().
new_from(Id) -> 
    Genotype     = edb:read(Id),
    Architecture = genotype:architecture(Genotype),
    AGroups      = genotype:actuator_groups(Genotype),
    SGroups      = genotype:sensor_groups(Genotype),
    Phenotype = #phenotype{
        id = ?PHENOTYPE_ID,
        network   = architecture:new_network(edb:read(Architecture)),
        actuators = [actuator:rand(G) || G <- edb:read(AGroups)],
        sensors   = [  sensor:rand(G) || G <- edb:read(SGroups)] 
    },
    edb:write(Phenotype),
    Phenotype#phenotype.id.


%%%===================================================================
%%% Phenotype callbacks 
%%%===================================================================

%%-------------------------------------------------------------------
%% @doc Returns the id of a new phenotype.
%% @end
%%-------------------------------------------------------------------
controller(Phenotype_Id) -> 
    Phenotype = edb:read(Phenotype_Id),
    enn:start(Phenotype#phenotype.network),
    State = #{
        cortex    => enn:cortex_pid(Phenotype#phenotype.network),
        actuators => edb:read(Phenotype#phenotype.actuators),
        sensors   => edb:read(Phenotype#phenotype.sensors)
    },
    enter_sensors(State).

%%-------------------------------------------------------------------
%% @doc Enter loops (change of state).
%% @end
%%-------------------------------------------------------------------
enter_sensors(State) -> 
    run_sensors([], ?SENSORS(State), State).

enter_predict(Signals, State) -> 
    Predictions = cortex:predict(?CORTEX(State), Signals),
    enter_actuators(Predictions, State).

enter_actuators(Predictions, State) -> 
    run_actuators([], Predictions, ?ACTUATORS(State), State).

enter_fit(Errors, State) -> 
    _BP_Errors = cortex:fit(?CORTEX(State), Errors),
    enter_sensors(State).

%%-------------------------------------------------------------------
%% @doc Run sensors.
%% @end
%%-------------------------------------------------------------------
run_sensors(Signals, [Sensor|Sx], State) ->
    case Sensor(State) of 
        {  ok, Signal, NextState} -> 
            run_sensors([Signal | Signals], Sx, NextState);
        {stop, Reason, NextState} -> 
            terminate(Reason, NextState)
    end;
run_sensors(Signals, [], State) -> 
    enter_predict(lists:reverse(Signals), State).

%%-------------------------------------------------------------------
%% @doc Run actuators.
%% @end
%%-------------------------------------------------------------------
run_actuators(Errors, [Prediction|Px], [Actuator|Ax], State) ->
    case Actuator(Prediction, State) of 
         {  ok,                NextState} -> 
            run_actuators([  0.0|Errors], Px, Ax, NextState);
         {  ok,  Error, Score, NextState} -> 
            eevo:score(self(), Score),
            run_actuators([Error|Errors], Px, Ax, NextState);
         {stop, Reason,        NextState} -> 
            terminate(Reason, NextState);
         {stop, Reason, Score, NextState} -> 
            eevo:score(self(), Score),
            terminate(Reason, NextState)
    end;
run_actuators(Errors, [], [], State) -> 
    enter_fit(lists:reverse(Errors), State).

%%-------------------------------------------------------------------
%% @doc Terminate.
%% @end
%%-------------------------------------------------------------------
terminate(Reason, _State) ->
    exit(Reason).


%%%===================================================================
%%% Internal functions
%%%===================================================================

File: lib/phenotype.ex
defmodule Phenotype do
  @moduledoc """
  """
  defstruct network: nil, actuators: [], sensors: []
  require Logger

  ### =================================================================
  ###  API
  ### =================================================================

  @doc """
  Returns the fields of an phenotype structure.
  """
  @spec fields() :: [atom()]
  def fields(), do: Map.from_struct(%Phenotype{}) |> Map.keys()

  @doc """
  Builds a phenotype from a genotype
  """
  @spec from(%Genotype{}) :: [%Phenotype{}]
  def from(%Genotype{} = genotype) do
    {:atomic, phenotype} =
      :mnesia.transaction(fn ->
        %Phenotype{
          network: :nnet.compile(genotype.model),
          actuators: select(genotype.actuators),
          sensors: select(genotype.sensors)
        }
      end)

    Logger.debug(what: "New phenotype", phenotype: phenotype)
    [phenotype]
  end

  @doc """
  Mutates the phenotype modifying the network, sensors and actuators
  """
  @spec mutate(%Phenotype{}) :: [%Phenotype{}]
  def mutate(%Phenotype{} = original) do
    mutation_function = fn ->
      %Phenotype{
        network: Network.mutate(original.network),
        actuators: for(x <- original.actuators, do: Actuator.mutate(x)),
        sensors: for(x <- original.sensors, do: Sensor.mutate(x))
      }
    end

    case :mnesia.transaction(mutation_function) do
      {:atomic, mutated} -> [mutated]
      {:aborted, _reasn} -> mutate(%Phenotype{} = original)
    end
  end

  ### =================================================================
  ### Phenotype callbacks
  ### =================================================================

  @spec controller(%Phenotype{}) :: :eevo.agent_return()
  @doc """
  Starts the phenotype loop
  """
  def controller(%Phenotype{} = phenotype) do
    {:atomic, network_info} = :mnesia.transaction(fn -> :nnet.info(phenotype.network) end)

    state = %{
      network: :enn.start_link(phenotype.network),
      net_info: network_info,
      score_f: score_factor(network_info),
      cortex: :enn.cortex(phenotype.network),
      actuators: for(x <- phenotype.actuators, do: Database.dirty_read!({:actuator, x})),
      sensors: for(x <- phenotype.sensors, do: Database.dirty_read!({:sensor, x})),
      data: %{}
    }

    Logger.debug(what: "Starting phenotype")
    {:next, &enter_sensors/1, [state]}
  end

  @doc """
  Gets the signals from the phenotype sensors
  """
  def enter_sensors(state) do
    Logger.debug(what: "Entering sensors")
    run_sensors([], state.sensors, state)
  end

  def run_sensors(signals, [sensor | sx], state) do
    case Sensor.run(sensor, state.data) do
      {:ok, signal, data} ->
        run_sensors([signal | signals], sx, %{state | data: data})

      {:stop, reason} ->
        terminate(reason, 0.0, state)
    end
  end

  def run_sensors(signals, [], state) do
    Logger.debug(what: "Exiting sensors", signals: signals)
    {:next, &enter_actuators/2, [signals, state]}
  end

  @doc """
  Converts the sensors signals into actuator inputs and executes the actuators
  """
  def enter_actuators(signals, state) do
    predictions = :cortex.predict(state.cortex, signals)
    Logger.debug(what: "Entering actuators", predictions: predictions)
    run_actuators([], 0.0, predictions, state.actuators, state)
  end

  def run_actuators(errs, s_acc, [pred | px], [actuator | ax], state) do
    case Actuator.run(actuator, pred, state.data) do
      {:ok, score, data} ->
        s_acc = score + s_acc
        run_actuators([0.0 | errs], s_acc, px, ax, %{state | data: data})

      {:ok, err, score, data} ->
        s_acc = score + s_acc
        run_actuators([err | errs], s_acc, px, ax, %{state | data: data})

      {:stop, reason} ->
        terminate(reason, s_acc, state)

      {:stop, reason, score} ->
        s_acc = score + s_acc
        terminate(reason, s_acc, state)
    end
  end

  def run_actuators(errors, score_acc, [], [], state) do
    _bp_errors = :cortex.fit(state.cortex, errors)
    score = score_acc * state.score_f
    Logger.debug(what: "Exiting actuators", errors: errors, score: score)
    {:next, &enter_sensors/1, [state], [{:score, score}]}
  end

  @doc """
  Terminates the phenotype
  """
  def terminate(reason, score_acc, state) do
    :ok = :enn.stop(state.network)
    score = score_acc * state.score_f
    Logger.debug(what: "Terminating agent", reason: reason, score: score)
    {:stop, reason, [{:score, score}]}
  end

  ### =================================================================
  ###  Internal functions
  ### =================================================================

  defp select(groups) when is_list(groups) do
    for name <- groups, do: select(name)
  end

  defp select(name) do
    group = Database.read!({:group, name})
    Enum.random(group.members)
  end

  @square_e 7.3890560989306495
  defp score_factor(network_info) do
    0.5 + 1 / :math.log(@square_e + network_info.size)
  end
end
